### PR TITLE
KEYCLOAK-13319 Use newest WebDriver/Selenium for the WebAuthn testing

### PR DESCRIPTION
**Pull request by mabartos**
JIRA: [KEYCLOAK-13319](https://issues.redhat.com/browse/KEYCLOAK-13319)

This is a prototype for the WebAuthn testing, which use features from Selenium 4. Virtual Authenticators work as expected in the WebAuthn tests.

@pskopek @Pepo48 Could you please take a look at it?

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticatorFactory.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticatorFactory.java:59]: I did not review whole PR in details, but is it possible to remove this change from this PR?

It was recently added on purpose due the security reasons (WebAuthn passwordless is considered "1st factor authenticator" and user is not supposed to be able to setup it during authentication unlike 2nd factor mechanisms).

Let me know if you want more details and we can find the corresponding discussion (I figured the commit message doesn't have JIRA number and hence might be a little bit complicated to find it...)

**Review by mposolda — CHANGES_REQUESTED** [review]
@mabartos Thanks for the work! I did not review in details, but just added a comment to WebAuthnPasswordlessFactory class

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticatorFactory.java
Comment (mabartos) [reply services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticatorFactory.java:59]: @mposolda Oh, you're right. I remember now something was discussed before. Thanks!

**Comment by mabartos** [comment]
I need to fix some additional tests.

**Review by mposolda — CHANGES_REQUESTED** [review]
@mabartos This is great work! Thanks for it! I have few more concerns and added some additional comments:

- Does this PR work in the pipeline already?

- Any update for HOW-TO-RUN.md ? Or is it expected that this does not work yet in the community? HOW-TO-RUN.md has some notes related to WebAuthn, which are surely not valid anymore with these changes, so some update to this document will be nice in this PR.

- It seems WebAuthnRegisterAndLoginTest tests only single configuration of WebAuthn authenticator? Is it a plan to extend this to run with multiple various WebAuthn configurations (EG. BLE, NFC and others...). Or is it a plan to do it as a follow-up?

**Comment by mabartos** [comment]
@mposolda Thank you very much for your comments!

- I'm trying to integrate those tests, at least temporarily, to the pipeline, but I'm quite struggling with those custom packages. I haven't used the GH packages before and I need to investigate it more. Did you try to build the module in testsuite? Were you able to fetch those packages?
- I'll update the `HOW-TO-RUN.md` file for WebAuthn tests. 
- There are missing a lot of test cases for WebAuthn to be considered as properly verified feature. I've created a few follow-up tasks for it and I resolved some, but I need to wait for this PR to be merged. Please, see epic [WebAuthn Automated testing](https://issues.redhat.com/browse/KEYCLOAK-13131).

**Comment by mposolda** [comment]
> @mposolda Thank you very much for your comments!
> 
>     * I'm trying to integrate those tests, at least temporarily, to the pipeline, but I'm quite struggling with those custom packages. I haven't used the GH packages before and I need to investigate it more. Did you try to build the module in testsuite? Were you able to fetch those packages?

No, I did not yet tried it. I guess that if it runs in pipeline, it can be considered that it works?

One thing, which I am not sure, is using these credentials to download your privately build packages. I don't have much experience with this approach and not 100% sure whether it is good practice. However if @miquelsi or @pdrozd are fine with this approach and approves this PR, I am fine as well :-)

> 
>     * I'll update the `HOW-TO-RUN.md` file for WebAuthn tests.

That will be nice, Thanks!

> 
>     * There are missing a lot of test cases for WebAuthn to be considered as properly verified feature. I've created a few follow-up tasks for it and I resolved some, but I need to wait for this PR to be merged. Please, see epic [WebAuthn Automated testing](https://issues.redhat.com/browse/KEYCLOAK-13131).

Will it be useful for you to have this PR merged to be able to continue with the follow-up work? From my point of view, this PR is ready to go once:
- Minor typo in BTE is fixed (See inline comment)
- Assume is removed (See inline comment)
- Pavel or Miquel approve this PR. Especially I am not sure about the part with the custom packages. I am adding them as reviewers to confirm whether they are ok with this approach.

**Comment by vanrar68** [comment]
Selenium 4 is now officially available. Apparently this version is required if we want to test WebAuthn LoginLess authentication scenarios (resident key and user verification set to 'required'). I'll add the tests for #7860 as soon as this is merged

**Comment by mabartos** [comment]
@vanrar68 I'm working on a WebAuthn testsuite, but it seems, there won't be any other major changes in this PR, so you can follow up on this PR. Moreover, probably we'll be able to properly execute those tests in a pipeline soon, therefore this PR will be merged soon.

**Comment by mabartos** [comment]
@mposolda Pipeline tests passed: https://keycloak-jenkins.com/job/universal-test-pipeline-server/1033/console

**Comment by mabartos** [comment]
The GH actions failures are not related to this PR.

**Comment by vanrar68** [comment]
Hi! it seems webauthn tests are failing on Windows platform because chrome 64bits webdriver are not available on remote storage (no matter which version of chrome is installed on the machine performing the tests). Google seems to publish 32bits driver exclusively for Windows platform. It used to work nicely with old webauthn test framework though:

`Caused by: org.jboss.arquillian.drone.webdriver.binary.downloading.source.MissingBinaryException: There wasn't found any binary with the key matching regex \Q96.0.4664.45/chromedriver_win64.zip\E in the storage: https://chromedriver.storage.googleapis.com/
	at org.jboss.arquillian.drone.webdriver.binary.downloading.source.XmlStorageSource.getReleaseForVersion(XmlStorageSource.java:157)
	at org.jboss.arquillian.drone.webdriver.binary.downloading.source.XmlStorageSource.getLatestRelease(XmlStorageSource.java:82)
	at org.jboss.arquillian.drone.webdriver.binary.downloading.source.XmlStorageSource.getLatestRelease(XmlStorageSource.java:75)
	at org.jboss.arquillian.drone.webdriver.binary.handler.AbstractBinaryHandler.downloadAndPrepare(AbstractBinaryHandler.java:175)
	at org.jboss.arquillian.drone.webdriver.binary.handler.AbstractBinaryHandler.checkAndSetBinary(AbstractBinaryHandler.java:61)
	... 12 more
`

**Comment by vanrar68** [comment]
> Hi! it seems webauthn tests are failing on Windows platform because chrome 64bits webdriver are not available on remote storage (no matter which version of chrome is installed on the machine performing the tests). Google seems to publish 32bits driver exclusively for Windows platform. It used to work nicely with old webauthn test framework though:

I managed to make the webauthn tests work on Windows platform by adding the -Dwebdriver.chrome.driver=C:/path/to/chromedriver.exe option to the maven command line. Could be interesting to add to the HOW-TO-RUN.md

**Comment by mabartos** [comment]
@vanrar68 Thanks for your information. I'll create a new issue for Arquillian Drone and note the workaround to the HOW-TO-RUN.md

**Review by mposolda — APPROVED** [review]
@mabartos Thanks for all your work on this PR!

@vanrar68 Thanks for the additional testing and verifications!

**Comment by mabartos** [comment]
@vanrar68 The issue for that was created. See https://github.com/arquillian/arquillian-extension-drone/issues/300

**Pull request merged**
